### PR TITLE
Add NatSpec to Style-Guide

### PR DESCRIPTION
**Pull request by 3esmit**
related to #2270

### CODE REVIEW

**Comment by 3esmit** [comment]
Need to review if code samples follow all style-guide and if sample code is valid for current solidity version.

**Comment by pirapira** [comment]
The appveyor failure is, I guess, due to infrastructural reasons.  So I created an issue https://github.com/ethereum/solidity/issues/2272 .

File: docs/layout-of-source-files.rst [hunk 1]
```diff
@@ -182,7 +182,7 @@ Single-line comments (``//``) and multi-line comments (``/*...*/``) are possible
 
 
 Additionally, there is another type of comment called a natspec comment,
-for which the documentation is not yet written. They are written with a
+for which is detailed in the style-guide. They are written with a
```
Comment (pirapira) [inline docs/layout-of-source-files.rst]: I guess `for` at the beginning of this line should be removed now.

File: docs/style-guide.rst [hunk 2]
```diff
@@ -750,3 +750,124 @@ General Recommendations
...
+      /// @param to The address of the recipient of the GavCoin
+      /// @param valueInmGav The GavCoin value to send
+      function send(address to, uint256 valueInmGAV) {
+        if (balances[msg.sender] >= valueInmGAV) {
+          balances[to] += valueInmGAV;
+          balances[msg.sender] -= valueInmGAV;
+        }
+      }
+
+      /// @notice `(balanceInmGAV / 1000).fixed(0,3)` GAV is the total funds available to `who.address()`.
```
Comment (pirapira) [inline docs/style-guide.rst]: Shall the `notice` field contain a period or not?

File: docs/style-guide.rst [hunk 3]
```diff
@@ -750,3 +750,124 @@ General Recommendations
...
+    + @title: This is a title that should describe the contract and go above the contract definition
+    + @author: The name of the author of the contract. Should also go above the contract definition.
+    + @notice: Represents user documentation. This is the text that will appear to the user to notify him
+      of what the function he is about to execute is doing
+    + @dev: Represents developer documentation. This is documentation that would only be visible to the
+      developer.
+    + @param: Documents a parameter just like in doxygen. Has to be followed by the parameter name.
+    + @return: Documents the return type of a contract's function.
+
+- If any of the above are missing they are simply considered as blank and it's not illegal to omit any of them
```
Comment (pirapira) [inline docs/style-guide.rst]: Maybe a period is missing.

File: docs/style-guide.rst [hunk 4]
```diff
@@ -750,3 +750,124 @@ General Recommendations
 =======================
 
 TODO
+
+***********
+NatSpec
```
Comment (axic) [inline docs/style-guide.rst]: I'd suggest to move this to its own file.

File: docs/style-guide.rst [hunk 5]
```diff
@@ -750,3 +750,124 @@ General Recommendations
...
+Solidity contracts can have a special form of comments that form the basis of the Ethereum Natural Specification Format. For a usage example please check [here](https://github.com/ethereum/wiki/wiki/Natspec-Example/).
+
+Documentation Example
+=====================
+
+Documentation is inserted above the function following the doxygen notation of either one or multiple lines starting with `///` or a multiline comment starting with `/**` and ending with `*/`.
+
+As an example consider the documentation of the following function::
+
+    /// @notice Send `(valueInmGAV / 1000).fixed(0,3)` GAV from the account of
```
Comment (axic) [inline docs/style-guide.rst]: Can this be changed into a generic token example?

File: docs/style-guide.rst [hunk 6]
```diff
@@ -750,3 +750,124 @@ General Recommendations
...
+    + @notice: Represents user documentation. This is the text that will appear to the user to notify him
+      of what the function he is about to execute is doing
+    + @dev: Represents developer documentation. This is documentation that would only be visible to the
+      developer.
+    + @param: Documents a parameter just like in doxygen. Has to be followed by the parameter name.
+    + @return: Documents the return type of a contract's function.
+
+- If any of the above are missing they are simply considered as blank and it's not illegal to omit any of them
+ 
+- `(valueInmGAV / 1000).fixed(0,3)` A dynamic expression. This should be a valid Javascript/Paperscript expression, which when evaluated in an EVM Javascript environment initialised with various system values (such as parameters).
```
Comment (axic) [inline docs/style-guide.rst]: What is paperscript?

File: docs/style-guide.rst [hunk 7]
```diff
@@ -750,3 +750,124 @@ General Recommendations
...
+
+User Documentation
+==================
+
+The above documentation will produce the following user documentation json file as output::
+
+    {
+      "source": "...",
+      "language": "Solidity",
+      "languageVersion": 1,
```
Comment (axic) [inline docs/style-guide.rst]: I think these are not part of the userdoc/devdoc.

**Comment by axic** [comment]
Solidity doesn't implement natspec as it is in the wiki, there is no `invariants` or `constructor` fields.

File: docs/style-guide.rst (hunk 5 above)
Comment (chriseth) [reply docs/style-guide.rst]: NatSpec was changed years ago to support javascript snippets and not some weird other formatting language.

**Comment by pirapira** [comment]
@3esmit are you still around?

**Comment by axic** [comment]
@pirapira this is assigned to me, I'm working on it

**Comment by pirapira** [comment]
@axic Now I see it.

**Comment by 3esmit** [comment]
@pirapira Sorry, I missed this PR, but I'm still around.

**Comment by ChrisChinchilla** [comment]
OK all, I am clearing out the documentation PRs and all of them are mine apart from this one.

@3esmit do you feel this is still current, and anything I can do to get it merged?

@axic Is this still a relevant feature to document?

**Comment by 3esmit** [comment]
Seems outdated. But the natspec docs still missing.

**Comment by ChrisChinchilla** [comment]
@axic @chriseth and @3esmit I updated and rebased. I also stripped this down to style recommendations instead of explanation. There may still be too much explanation, but actually, I think the docs are missing some of the explanation, but that's not for this PR.

File: docs/style-guide.rst [hunk 8]
```diff
@@ -1093,3 +1093,64 @@ General Recommendations
...
+        /// Store `x`
+        /// @param x the new value to store
+        /// @dev stores the number in the state variable `storedData`.
+        function set(uint x) public {
+            storedData = x;
+        }
+
+        /// Return the stored value.
+        //// @dev retrieves the value of the state variable `storedData`
+        //// @return the stored value
```
Comment (leonardoalt) [inline docs/style-guide.rst]: Why do these have 4 `/`?

File: docs/style-guide.rst [hunk 9]
```diff
@@ -1093,3 +1093,64 @@ General Recommendations
...
+.. _natspec:
+
+*******
+NatSpec
+*******
+
+Solidity contracts can have a form of comments that are the basis of the
+Ethereum Natural Language Specification Format.
+
+Add the comments above the function or contract following `doxygen <http://www.doxygen.nl>`_ notation
```
Comment (leonardoalt) [inline docs/style-guide.rst]:
```suggestion
Add comments above functions or contracts following `doxygen <http://www.doxygen.nl>`_ notation
```

File: docs/style-guide.rst [hunk 10]
```diff
@@ -1093,3 +1093,64 @@ General Recommendations
...
+of one or multiple lines starting with `///` or a
+multiline comment starting with `/**` and ending with `*/`.
+
+For example, the contract from `a simple smart contract <simple-smart-contract>`_ with the comments
+added looks like the below::
```
Comment (leonardoalt) [inline docs/style-guide.rst]:
```suggestion
added looks like the one below::
```
To me `the below` sounded a bit weird, not sure though

File: docs/style-guide.rst [hunk 11]
```diff
@@ -1093,3 +1093,64 @@ General Recommendations
...
+    pragma solidity >=0.4.0 <0.6.0;
+
+    /// @author The Solidity Team
+    /// @title A simple storage example
+    contract SimpleStorage {
+        uint storedData;
+
+        /// Store `x`
+        /// @param x the new value to store
+        /// @dev stores the number in the state variable `storedData`.
```
Comment (leonardoalt) [inline docs/style-guide.rst]:
```suggestion
        /// @dev stores the number in the state variable `storedData`
```

File: docs/style-guide.rst [hunk 12]
```diff
@@ -1093,3 +1093,64 @@ General Recommendations
...
+users of the contract who have never seen the source code, so it should make
+as little assumptions about the inner details as possible.
+All tags are optional.
+
++-------------+-------------------------------------------+-------------------------------+
+| Tag         | Description                               | Context                       |
++=============+===========================================+===============================+
+| ``@title``  | A title that describes the contract       | contract, interface           |
++-------------+-------------------------------------------+-------------------------------+
+| ``@author`` | The name of the author of the contract    | contract, interface, function |
```
Comment (leonardoalt) [inline docs/style-guide.rst]:
```suggestion
| ``@author`` | The name of the author   | contract, interface, function |
```
Also applies to functions

File: docs/style-guide.rst [hunk 13]
```diff
@@ -1093,3 +1093,64 @@ General Recommendations
...
++-------------+-------------------------------------------+-------------------------------+
+| ``@notice`` | Explanation of what a function does       | contract, interface, function |
```
Comment (leonardoalt) [inline docs/style-guide.rst]:
```suggestion
| ``@notice`` | Explanation of a functionality       | contract, interface, function |
```

**Review by leonardoalt — DISMISSED** [review]

**Pull request merged**